docs/hlo_dumps.md

```diff
@@ -0,0 +1,219 @@
+
+## Local Execution
+
+### Using Environment Variables
+
+You can set the `XLA_FLAGS` environment variable with the necessary flags to get dumps. This works for JAX, TensorFlow, and PyTorch/XLA.
+
+To dump HLO modules and other debugging information to a specific directory, run your program with the `--xla_dump_to` flag:
+
+```shell
+XLA_FLAGS="--xla_dump_to=DIRECTORY_PATH"
+```
+
+For example, you can use `/tmp` or `/tmp/xladump` as the paths.
+
+By default, this dumps HLO modules as text, at the very beginning and end of the optimization pipeline.
+
+You can also explicitly specify the format:
+
+1. Text dumps
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_as_text --xla_dump_to=DIRECTORY_PATH"
+```
+
+2. HLO protos
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_as_proto --xla_dump_to=DIRECTORY_PATH"
+```
+
+3. HLO Snapshots
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_snapshots --xla_dump_to=DIRECTORY_PATH"
+```
+
+4. Graph render with graphviz server (only works well for small graphs)
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_as_url --xla_dump_to=DIRECTORY_PATH"
+```
+
+5. Graph render to HTML file (only works well for small graphs)
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_as_html --xla_dump_to=DIRECTORY_PATH"
+```
+
+For larger graphs, you can use `interactive_graphviz` to visualize parts of the graph.
+
+**Note:** If `--xla_dump_to` is not specified but another dumping flag is specified, it will dump to stdout. But the dump will not include binary data, e.g., proto files, to stdout.
+
+## Dump Specific Intermediate Passes
+
+In addition to the standard pre-optimized / final-optimized HLOs, you can also dump the state of HLOs after a particular compiler pass.
+
+```shell
+XLA_FLAGS="--xla_dump_hlo_pass_re=regex --xla_dump_to=DIRECTORY_PATH"
+```
+
+HLO modules will be dumped for the passes whose names match the regular expression (regex). For example, you can observe the HLOs resulting from passes related to SPMD partitioning with:
+
+```shell
+XLA_FLAGS="--xla_dump_to=DIRECTORY_PATH --xla_dump_hlo_pass_re=spmd|propagation"
+```
+
+To dump the result after every XLA pass (this will result in a lot of files), you can set:
+
+```shell
+XLA_FLAGS="--xla_dump_to=DIRECTORY_PATH --xla_dump_hlo_pass_re=.*"
+```
+
+### JAX-specific Options
+
+#### Programmatically in JAX
+
+Instead of passing flags or environment variables, you can also programmatically dump HLO using JAX’s `lower` and `compil`e APIs.
+
+Locally fetch the unoptimized original lowered HLO with:
+
+```python
+jax.jit(f).lower(*args).as_text('hlo')
+```
+
+For dumping to files during HLO compilation passes, specify:
+
+```python
+compilation_args = {
+    'xla_dump_to': DIRECTORY_PATH,
+    'xla_dump_hlo_pass_re': 'spmd|propagation', # or some other pass filter
+    ...
+    }
+
+jax.jit(f).lower(*args).compile(compilation_args)
+```
+
+#### Dump jaxprs
+
+`jaxpr`s are JAX's intermediate representation for program traces. To dump this, set the environment variables:
```

Comment:
Maybe link to JAX's jaxpr documentation? https://docs.jax.dev/en/latest/jaxpr.html